core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,281 +19,590 @@
-     *
-     * @return The map of parameters the script will be executed with.
+     * Constructor for simple script using the default language and default type.
+     * @param idOrCode The id or code to use dependent on the default script type.
      */
-    public Map<String, Object> getParams() {
-        return params;
+    public Script(String idOrCode) {
+        this(DEFAULT_SCRIPT_TYPE, DEFAULT_SCRIPT_LANG, idOrCode, Collections.emptyMap(), Collections.emptyMap());
     }
 
     /**
-     * @return The content type of the script if it is an inline script and the script has been defined as json
-     *         or yaml content instead of a plain string.
+     * Constructor for a script that does not need to use compiler options.
+     * @param type     The {@link ScriptType}.
+     * @param lang     The lang for this {@link Script}.
+     * @param idOrCode The id for this {@link Script} if the {@link ScriptType} is {@link ScriptType#FILE} or {@link ScriptType#STORED}.
+     *                 The code for this {@link Script} if the {@link ScriptType} is {@link ScriptType#INLINE}.
+     * @param params   The user-defined params to be bound for script execution.
      */
-    public XContentType getContentType() {
-        return contentType;
+    public Script(ScriptType type, String lang, String idOrCode, Map<String, Object> params) {
+        this(type, lang, idOrCode, Collections.emptyMap(), params);
     }
 
+    /**
+     * Constructor for a script that requires the use of compiler options.
+     * @param type     The {@link ScriptType}.
+     * @param lang     The lang for this {@link Script}.
+     * @param idOrCode The id for this {@link Script} if the {@link ScriptType} is {@link ScriptType#FILE} or {@link ScriptType#STORED}.
+     *                 The code for this {@link Script} if the {@link ScriptType} is {@link ScriptType#INLINE}.
+     * @param options  The options to be passed to the compiler for use at compile-time.
+     * @param params   The user-defined params to be bound for script execution.
+     */
+    public Script(ScriptType type, String lang, String idOrCode, Map<String, String> options, Map<String, Object> params) {
+        this.idOrCode = Objects.requireNonNull(idOrCode);
+        this.type = Objects.requireNonNull(type);
+        this.lang = Objects.requireNonNull(lang);
+        this.options = Collections.unmodifiableMap(Objects.requireNonNull(options));
+        this.params = Collections.unmodifiableMap(Objects.requireNonNull(params));
+
+        if (type != ScriptType.INLINE && !options.isEmpty()) {
+            throw new IllegalArgumentException(
+                "Compiler options [" + options + "] cannot be specified at runtime for [" + type + "] scripts.");
+        }
+    }
+
+    /**
+     * Creates a {@link Script} read from an input stream.
+     */
+    public Script(StreamInput in) throws IOException {
+        // Version 6.0 requires all Script members to be non-null and supports the potential
```

Comment:
5.1+ I think, actually.

Have a look at `VersionTests#testUnknownVersions` and the note right underneath `Version#CURRENT`.